### PR TITLE
Zero3 Fix allreduce optimization for extra large tensor

### PR DESCRIPTION
**Pull request by hablb**
Grad tensors that don't fit in the bucket flat buffer are not added to it, but still added to params_in_ipg_bucket if such tensors exists use reduce_scatter of params_in_ipg_bucket instead of allreduce. since allreduce assumes all grads are in ipg_bucket_flat_buffer.

Add test for reduce scatter=false
Fix padding to zeros instead of undefined values

### CODE REVIEW

File: deepspeed/runtime/zero/stage3.py [hunk 1]
```diff
@@ -1167,6 +1167,7 @@ def __avg_scatter_contiguous_grads(self, buffer_to_reduce: Tensor) -> List[Tenso
                 padded_partition = torch.empty(param.partition_numel(), device=grad.device, dtype=grad.dtype)
                 if partition.numel() > 0:
                     padded_partition[:partition.numel()] = partition
+                padded_partition[partition.numel():] = 0
```
[the remark this answers is not in the document]
Comment (hablb) [reply deepspeed/runtime/zero/stage3.py:1170]: Yes.
Calling torch.zeros in line 1167 will cause the whole bucket be filled with zeros while most of them will be overwritten in line 1169.
Doing it this way will insert zeros only at the padding if exists.
Would you rather I change 1167 to torch.zeros?

File: deepspeed/runtime/zero/stage3.py [hunk 2]
```diff
@@ -1118,7 +1118,7 @@ def __reduce_and_partition_ipg_grads(self, safe_mode: bool = False) -> None:
             if safe_mode:
                 assert_ints_same_as_other_ranks([p.ds_id for p in self.params_in_ipg_bucket])
 
-            if self.contiguous_gradients and not self.reduce_scatter:
+            if self.contiguous_gradients and self.elements_in_ipg_bucket <= self.reduce_bucket_size and not self.reduce_scatter:
```
[the remark this answers is not in the document]
Comment (hablb) [reply deepspeed/runtime/zero/stage3.py:1121]: Correct. I encountered it when I set reduce_bucket_size to a small number causing some grads to exceed it's size.

File: deepspeed/runtime/zero/stage3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjruwase) [reply deepspeed/runtime/zero/stage3.py:1170]: Yes, I suggest changing 1167 to `torch.zeros`, so 1170 is not needed anymore.

[the remark this answers is not in the document]
Comment (hablb) [reply deepspeed/runtime/zero/stage3.py]: Done

**Review by tjruwase — APPROVED** [review]
